src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+// See the LICENSE file in the project root for more information.
+
+#pragma warning disable 1591
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a buffer that can read a sequential series of bytes.
+    /// </summary>
+    public readonly partial struct ReadOnlyBuffer<T>
+    {
+        private const int IndexBitMask = 0x7FFFFFFF;
+
+        internal const int MemoryListStartMask = 0;
+        internal const int MemoryListEndMask = 0;
+
+        internal const int ArrayStartMask = 0;
+        internal const int ArrayEndMask = 1 << 31;
+
+        internal const int OwnedMemoryStartMask = 1 << 31;
+        internal const int OwnedMemoryEndMask = 0;
+
+        internal readonly SequencePosition BufferStart;
+        internal readonly SequencePosition BufferEnd;
+
+        public static readonly ReadOnlyBuffer<T> Empty = new ReadOnlyBuffer<T>(new T[0]);
+
+        /// <summary>
+        /// Length of the <see cref="ReadOnlyBuffer{T}"/> in bytes.
+        /// </summary>
+        public long Length => GetLength(BufferStart, BufferEnd);
+
+        /// <summary>
+        /// Determines if the <see cref="ReadOnlyBuffer{T}"/> is empty.
+        /// </summary>
+        public bool IsEmpty => Length == 0;
+
+        /// <summary>
+        /// Determins if the <see cref="ReadOnlyBuffer{T}"/> is a single <see cref="Memory{Byte}"/>.
+        /// </summary>
+        public bool IsSingleSegment => BufferStart.Segment == BufferEnd.Segment;
+
+        public ReadOnlyMemory<T> First
+        {
+            get
+            {
+                TryGetBuffer(BufferStart, BufferEnd, out var first, out _);
+                return first;
+            }
+        }
+
+        /// <summary>
+        /// A position to the start of the <see cref="ReadOnlyBuffer{T}"/>.
+        /// </summary>
+        public SequencePosition Start => BufferStart;
+
+        /// <summary>
+        /// A position to the end of the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public SequencePosition End => BufferEnd;
+
+        private ReadOnlyBuffer(object startSegment, int startIndex, object endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+
+            BufferStart = new SequencePosition(startSegment, startIndex);
+            BufferEnd = new SequencePosition(endSegment, endIndex);
+        }
+
+        public ReadOnlyBuffer(IMemoryList<T> startSegment, int startIndex, IMemoryList<T> endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+            Debug.Assert(startSegment.Memory.Length >= startIndex);
+            Debug.Assert(endSegment.Memory.Length >= endIndex);
+
+            BufferStart = new SequencePosition(startSegment, startIndex | MemoryListStartMask);
```

Comment:
MemoryListStartMask / MemoryListEndMask are both 0. The bitwise or here is redundant.

Reply:
It is redundant but kept for consistency and to show the pattern in case constant values would change.

Reply:
Fair enough. As long as it gets optimized away by the JIT.